### PR TITLE
Users/ersciple/m108findandmatch

### PR DESCRIPTION
**Pull request by ericsciple**

### CODE REVIEW

File: node/task.ts [hunk 1]
```diff
@@ -1094,6 +1095,434 @@ class FindItem {
...
+
+            pattern = pattern.substring(negateCount); // trim leading '!'
+        }
+
+        let isIncludePattern = negateCount == 0 ||
+            (negateCount % 2 == 0 && !matchOptions.flipNegate) ||
+            (negateCount % 2 == 1 && matchOptions.flipNegate);
+
+        // set nonegate - brace expansion could result in a leading '!'
+        matchOptions.nonegate = true;
```
Comment (TingluoHuang) [inline node/task.ts]: set matchOptions.flipNegate=false

File: node/task.ts [hunk 2]
```diff
@@ -1094,6 +1095,434 @@ class FindItem {
...
+
+        // convert slashes on Windows - unfortunately this means several things cannot be escaped on Windows,
+        // i.e. leading '#', brace expansion, extended globbing, and '[...]' globbing. this limitation is
+        // consistent with current limitations of minimatch (3.0.3).
+        //
+        // note, code below (2 places) relies on the assumption that '\' has been converted to '/' on Windows
+        pattern = process.platform == 'win32' ? pattern.replace(/\\/g, '/') : pattern;
+
+        // expand braces - required to accurately interpret findPath
+        let expanded: string[] = matchOptions.nobrace ? [ pattern ] : braceexpansion(pattern);
```
Comment (TingluoHuang) [inline node/task.ts]: check whether minimatch has minimatch.braceexpansion() function.

File: node/task.ts [hunk 3]
```diff
@@ -1094,6 +1095,434 @@ class FindItem {
...
+        for (let pattern of expanded) {
+            if (!pattern) {
+                continue;
+            }
+
+            if (isIncludePattern) {
+                // determine the findPath
+                let findPath = '';
+                let statOnly = false;
+                if (matchOptions.matchBase && !isRooted(pattern) && pattern.indexOf(path.sep) < 0) { // note, relies on assumption '\' have been converted to '/' on Windows
```
Comment (TingluoHuang) [inline node/task.ts]: path.sep or path.altsep

**Comment by ericsciple** [comment]
is findMatch() a better name?

**Comment by ericsciple** [comment]
@bryanmacfarlane this is ready now.

i went ahead and removed the glob functions and bumped the version to 2.0.0. i can revert removing the glob() functions if you think that's better at this point.

i also removed some gulp dependencies that dont appear to be used anymore.

i also changed the function name to findMatch (removed "and")

i want to followup with a small markdown discussing the match functions, the two typical usage patterns (i.e. filepath only vs root + text area), recommended default options, and touch lightly on the globbing syntaxes.

i've also started looking into porting to ps3 - will be a separate PR. i'm hopeful it will go fairly quickly.

**Comment by ericsciple** [comment]
also, fyi, i reviewed this again today with @TingluoHuang. he and i discussed it extensively at the beginning of the week since then i had integrated good feedback

**Comment by ericsciple** [comment]
i found two small issues i need to address before merging (didnt fallback to system.defaultworkingdirectory correctly, and forgot to delete glob functions although removed npm references)

**Pull request merged**